rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+                UploadState::CreatingUpload(_)
+                | UploadState::PuttingSingle(_)
+                | UploadState::Completing(_) => return Poll::Pending,
+                UploadState::Pending => {
+                    // If we didn't start a multipart upload, we can just do a single put.
+                    let part = Bytes::from(std::mem::take(&mut mut_self.buffer));
+                    let path = mut_self.path.clone();
+                    let store = mut_self.store.clone();
+                    let fut = Box::pin(async move {
+                        store.put(&path, part).await?;
+                        Ok(())
+                    });
+                    self.state = UploadState::PuttingSingle(fut);
+                }
+                UploadState::InProgress {
+                    futures,
+                    part_ids,
+                    multipart_id,
+                    part_idx,
+                } => {
+                    // Flush final batch
+                    if !mut_self.buffer.is_empty() && futures.len() < 10 {
+                        futures.push(Self::put_next_part(
+                            &mut_self.path,
+                            &mut_self.store,
+                            &mut mut_self.buffer,
+                            part_idx,
+                            multipart_id,
+                        ));
+                        // We need to go back to beginning of loop to poll the
+                        // new feature and get the waker registered on the ctx.
+                        continue;
+                    }
+
+                    // We handle the transition from in progress to completing here.
+                    if futures.is_empty() {
+                        let part_ids = std::mem::take(part_ids)
+                            .into_iter()
+                            .map(|maybe_id| {
+                                maybe_id.ok_or_else(|| {
+                                    io::Error::new(io::ErrorKind::Other, "missing part id")
+                                })
+                            })
+                            .collect::<io::Result<Vec<_>>>()?;
+                        let path = mut_self.path.clone();
+                        let store = mut_self.store.clone();
+                        let multipart_id = multipart_id.clone();
+                        let fut = Box::pin(async move {
+                            store
+                                .complete_multipart(&path, &multipart_id, part_ids)
+                                .await?;
+                            Ok(())
+                        });
+                        self.state = UploadState::Completing(fut);
+                    } else {
+                        return Poll::Pending;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Returned error from trying to upload a part.
+/// Has the part_idx and buffer so we can
```

Comment:
```suggestion
/// Has the part_idx and buffer so we can
/// print meaningful error messages
```

Although I'm not sure `part_idx` or `buffer` will be all that useful in an error message so maybe I'm missing something

Reply:
Ha actually this was meant so that we can re-use these for the retry logic. Updating the comment.